torchgeo/models/changestar.py

```diff
@@ -0,0 +1,229 @@
+                semantic segmentation model without semantic segmentation head.
+            seg_classifier: semantic segmentation head, typically a convolutional layer
+                followed by an upsampling layer.
+            changemixin: :class:`torchgeo.models.ChangeMixin` module
+            inference_mode: name of inference mode ``'t1t2'`` | ``'t2t1'`` | ``'mean'``.
+                ``'t1t2'``: concatenate bitemporal features in the order of t1->t2;
+                ``'t2t1'``: concatenate bitemporal features in the order of t2->t1;
+                ``'mean'``: the weighted mean of the output of ``'t1t2'`` and ``'t1t2'``
+        """
+        super().__init__()  # type: ignore[no-untyped-call]
+        self.dense_feature_extractor = dense_feature_extractor
+        self.seg_classifier = seg_classifier
+        self.changemixin = changemixin
+
+        if inference_mode not in ["t1t2", "t2t1", "mean"]:
+            raise ValueError(f"Unknown inference_mode: {inference_mode}")
+        self.inference_mode = inference_mode
+
+    def forward(self, x: Tensor) -> Dict[str, Any]:
+        """Forward pass of the model.
+
+        Args:
+            x: a bitemporal input tensor of shape [B, T, C, H, W]
+
+        Returns:
+            a dictionary containing
+                if training stage, returning
+                    bi_seg_logit: bitemporal semantic segmentation logit
+                    bi_change_logit: bidirected binary change detection logit
+                if inference stage, returning
+                    bi_seg_logit: bitemporal semantic segmentation logit
+                    change_prob:  binary change detection probability
+        """
+        b, t, c, h, w = x.shape
+        x = rearrange(x, "b t c h w -> (b t) c h w")
+        # feature extraction
+        bi_feature = self.dense_feature_extractor(x)
+        # semantic segmentation
+        bi_seg_logit = self.seg_classifier(bi_feature)
+        bi_seg_logit = rearrange(bi_seg_logit, "(b t) c h w -> b t c h w", t=t)
+
+        bi_feature = rearrange(bi_feature, "(b t) c h w -> b t c h w", t=t)
+        # change detection
+        c12, c21 = self.changemixin(bi_feature)
+
+        results: Dict[str, Any] = {}
+        if not self.training:
+            results.update({"bi_seg_logit": bi_seg_logit})
+            if self.inference_mode == "t1t2":
+                results.update({"change_prob": c12.sigmoid()})
+            elif self.inference_mode == "t2t1":
+                results.update({"change_prob": c21.sigmoid()})
+            elif self.inference_mode == "mean":
+                results.update(
+                    {
+                        "change_prob": torch.stack([c12, c21], dim=0)
+                        .sigmoid_()
+                        .mean(dim=0)
+                    }
+                )
+        else:
+            results.update(
+                {"bi_seg_logit": bi_seg_logit, "bi_change_logit": [c12, c21]}
```

Comment:
This is the only thing preventing the method signature from being `Dict[str, Tensor]`, can `c12` and `c21` be combined, or split into different dict keys?

Reply:
We could either torch.stack([c12, c21]) the tensors together or just make the type: Dict[str, Union[Tensor, List[Tensor]]. I think either is ok but whatever we choose should be explicitly stated in the docstring so users don't have to go to the source to understand what's happening.

Reply:
Thanks, revised.

Reply:
Lgtm